week-1-ts-ecosystem/w01e04-light-me-up/src/main.ts

```diff
@@ -8,14 +8,82 @@
  * Dodatkowo — rzucać błąd, jeśli zapas mocy się wyczerpie.
  * */
 
+const MILLISECONDS_IN_SECOND = 1000;
+
 class PowerSource {
-  private energySupply = 100
+  private energySupply = 100 // J
+
+  public consume(energy: number) {
+    const rest = this.energySupply - energy;
+
+    if (rest < 0) {
+      throw Error("Not enough energy");
+    }
 
-  consume(energy) {
-    this.energySupply -= energy
+    console.log(`Consume ${energy}J. Remaining energy: ` + rest + "J");
+    this.energySupply = rest;
   }
 }
 
 class LightBulb {
-  protected readonly powerConsumption = 20
+  protected readonly powerConsumption = 20 // W
+  private powerSource: PowerSource;
+  private consumption: number = -1;
+
+  constructor(powerSource: PowerSource) {
+    this.powerSource = powerSource;
+  }
+
+  consume () {
+    try {
+      this.powerSource.consume(this.powerConsumption);
+    } catch (error) {
+      console.log("Not enough energy. Switching off...");
+      this.switchOff();
+    }
+  }
+
+  switchOff () {
+      clearInterval(this.consumption);
+  }
+
+  switchOn (seconds: number) {
+      let secondsRemain = seconds;
+
+      this.consumption = setInterval(() => {
+        if (secondsRemain-- > 0) {
+          this.consume();
```

Comment:
Tutaj również zauważ, że Twoja żarówka - _zaświeci_ dopiero po `1s` od "Wkręcenia".

Nie jest to błąd w kodzie - tylko pod rozwagę do wzięcia przy dowożeniu funkcjonalności 
(czy tak może być?)